LoopKit/UserAlert.swift

```diff
@@ -0,0 +1,108 @@
+}
+
+/// Protocol that describes something that can deal with a user's response to an alert.
+public protocol UserAlertResponder {
+    /// Acknowledge alerts with a given type identifier
+    func acknowledgeAlert(typeIdentifier: UserAlert.TypeIdentifier) -> Void
+}
+
+/// Structure that represents an Alert that needs to be shown to the User.
+public struct UserAlert {
+    /// Representation of an alert Trigger
+    public enum Trigger {
+        /// Trigger the alert immediately
+        case immediate
+        /// Delay triggering the alert by `interval`, but issue it only once.
+        case delayed(interval: TimeInterval)
+        /// Delay triggering the alert by `repeatInterval`, and repeat at that interval until cancelled or unscheduled.
+        case repeating(repeatInterval: TimeInterval)
+    }
+    /// Content of the alert, either for foreground or background alerts
+    public struct Content {
+        public let title: String
+        public let body: String
+        /// Should this alert be deemed "critical" for the User?  Handlers will determine how that is manifested.
+        public let isCritical: Bool
+        // TODO: when we have more complicated actions.  For now, all we have is "acknowledge".
+//        let actions: [UserAlertAction]
+        public let acknowledgeActionButtonLabel: String
+        public init(title: String, body: String, acknowledgeActionButtonLabel: String, isCritical: Bool = false) {
+            self.title = title
+            self.body = body
+            self.acknowledgeActionButtonLabel = acknowledgeActionButtonLabel
+            self.isCritical = isCritical
+        }
+    }
+    public struct Identifier: Equatable {
+        /// Unique device manager identifier from whence the alert came, and to which alert acknowledgements should be directed.
+        public let deviceManagerInstanceIdentifier: String
+        /// Per-alert unique identifier, for instance to group alert types.  This is the identifier that will be used to acknowledge the alert.
+        public let typeIdentifier: TypeIdentifier
+        public init(deviceManagerInstanceIdentifier: String, typeIdentifier: TypeIdentifier) {
+            self.deviceManagerInstanceIdentifier = deviceManagerInstanceIdentifier
+            self.typeIdentifier = typeIdentifier
+        }
+        /// An opaque value for this tuple for unique identification of the alert across devices.
+        public var value: String {
+            return "\(deviceManagerInstanceIdentifier).\(typeIdentifier)"
+        }
+    }
+    /// This type represents a per-alert-type identifier, but not necessarily unique across devices.  Each device may have its own Swift type for this,
+    /// so conversion to String is the most convenient, but aliasing the type is helpful because it is not just "any String".
+    public typealias TypeIdentifier = String
+    public typealias AcknowledgeCompletion = (TypeIdentifier) -> Void
```

Comment:
Does this support a different use from UserAlertResponder.acknowledgeAlert() ?  Also, supporting persistence could be hard for this since it's being used below in a closure.

Reply:
Yeah, I went back and forth on this.  I think I'll get rid of this in favor of AlertResponder.